core/src/main/java/org/mskcc/cbio/portal/servlet/PathologyReportView.java

```diff
@@ -0,0 +1,224 @@
+        // Convert from the expected url path 'studyId/patientId/sampleId.pdf' to the expected filesystem path 'studyId.patientId.sampleId.pdf'
+        final String convertedPath = getRequestedPath(request).replace("/", ".");
+        return new File(DATA_DIRECTORY, convertedPath);
+    }
+    
+    private void setError(HttpServletRequest request, int httpStatusCode, String errorMessage) {
+        request.setAttribute(ERROR_CODE, Integer.toString(httpStatusCode));
+        request.setAttribute(ERROR_MSG, errorMessage);
+    }
+    
+    private boolean validateRequest(HttpServletRequest request) throws IOException, DaoException {
+        final String requestedPath = getRequestedPath(request);
+        final String[] path = requestedPath.split("/");
+        if (path.length != 3) {
+            setError(request, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Unable to process requested pathology " +
+                "report path. Ensure it is in the format cbioportal/pathology_report/study_id/patient_id/sample_id.pdf");
+            return false;
+        }
+
+        final String cancerStudyId = path[0];
+        CancerStudy cancerStudy = DaoCancerStudy.getCancerStudyByStableId(cancerStudyId);
+        if (cancerStudy == null) {
+            setError(request, HttpServletResponse.SC_NOT_FOUND, "No such cancer study with id '" + cancerStudyId + "'.");
+            return false;
+        }
+        if (accessControl.isAccessibleCancerStudy(cancerStudyId).size() != 1) {
+            setError(request, HttpServletResponse.SC_FORBIDDEN, 
+                "You are not authorized to access the cancer study with id '" + cancerStudyId + "'.");
+            return false;
+        }
+        
+        final String patientId = path[1];
+        Patient patient = DaoPatient.getPatientByCancerStudyAndPatientId(cancerStudy.getInternalId(), patientId);
+        if (patient == null) {
+            setError(request, HttpServletResponse.SC_NOT_FOUND, "No such patient with id '" + patientId + "' " +
+                "within the cancer study with id '" + cancerStudyId + "'");
+            return false;
+        }
+        
+        final String sampleId = FilenameUtils.removeExtension(path[2]);
+        Sample sample = DaoSample.getSampleByCancerStudyAndSampleId(cancerStudy.getInternalId(), sampleId, false);
+        if (sample == null) {
+            setError(request, HttpServletResponse.SC_NOT_FOUND, "No such sample with id '" + sampleId + "' " +
+                "within the cancer study with id '" + cancerStudyId + "'");
+            return false;
+        }
+        
+        if (DATA_DIRECTORY == null) {
+            setError(request, HttpServletResponse.SC_NOT_FOUND, "The internal location of pathology reports is undefined");
+            return false;
+        }
+        File requestedFile = getRequestedFile(request);
+        if (!requestedFile.exists() || requestedFile.isDirectory()) {
```

Comment:
This check (and the others) are performed within this function to be able to respond with an appropriate HTTP response status code and message. The serve file function does throw an `IO exception` which should get raised if something goes wrong during reading (such as if the file doesn't exist or is a directory), but that doesn't get captured/handled anywhere. If an issue during reading did occur and the `IO exception` is thrown, that would then get re-thrown by this class `PathologyReportView` since it extends `HttpServlet` which also throws an `IO exception` for the `processRequest` method.